stairway/src/main/java/bio/terra/stairway/impl/StairwayThreadPool.java

```diff
@@ -29,10 +33,45 @@ int getQueuedFlights() {
     return getQueue().size();
   }
 
+  /**
+   * Submits a Runnable task for execution and returns a Future representing that task. The Future's
+   * get method will return null upon successful completion.
+   *
+   * @param flightRunner the flight to submit
+   * @param flightContext
+   * @return
+   */
+  protected Future<?> submitWithMdcAndFlightContext(
+      Runnable flightRunner, FlightContext flightContext) {
+    // Save the calling thread's context before potentially submitting to a child thread
+    Map<String, String> callingThreadContext = MDC.getCopyOfContextMap();
+    Runnable flightRunnerWithMdc =
+        () -> {
+          Map<String, String> initialContext = MDC.getCopyOfContextMap();
+          initializeFlightMdc(callingThreadContext, flightContext);
+          try {
+            flightRunner.run();
+          } finally {
+            MdcUtils.overwriteContext(initialContext);
+          }
+        };
+    return super.submit(flightRunnerWithMdc);
+  }
+
+  private void initializeFlightMdc(
+      Map<String, String> callingThreadContext, FlightContext flightContext) {
+    // Any leftover context on the thread will be fully overwritten:
+    MdcUtils.overwriteContext(callingThreadContext);
+    // If the calling thread's context contains flight and step context from a parent flight, this
+    // will be overwritten below:
+    MdcUtils.addFlightContextToMdc(flightContext);
+    MdcUtils.removeStepContextFromMdc(flightContext);
+  }
+
   protected void beforeExecute(Thread t, Runnable r) {
-    super.beforeExecute(t, r);
     int active = activeTasks.incrementAndGet();
     logger.debug("before: " + active);
+    super.beforeExecute(t, r);
```

Comment:
Reading the docstring on this method, it's recommended to call the super at the end to allow for potential nesting.

(I also read the docstring on `afterExecute`, and the super call at the beginning matches the recommendation.)